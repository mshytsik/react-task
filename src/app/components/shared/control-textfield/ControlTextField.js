import { FormControl, TextField } from "@mui/material";
import { Controller } from "react-hook-form";

const ControlTextField = ({
  control,
  name,
  label,
  isRequired,
  maxLength,
  helperText = "",
  error,
}) => {
  const rules = { required: isRequired };
  if (maxLength) {
    rules.maxLength = maxLength;
  }

  const isTextarea = maxLength > 1000;

  return (
    <FormControl fullWidth sx={{ mb: "20px" }}>
      <Controller
        control={control}
        name={name}
        rules={rules}
        render={({ field }) => (
          <TextField
            {...field}
            label={label}
            size="medium"
            required={isRequired}
            error={Boolean(error)}
            helperText={
              error?.type === "maxLength"
                ? `Maximum length is ${maxLength} symbols.`
                : helperText
            }
            multiline={isTextarea}
            minRows={isTextarea ? 4 : undefined}
            maxRows={isTextarea ? 10 : undefined}
          />
        )}
      />
    </FormControl>
  );
};

export default ControlTextField;
