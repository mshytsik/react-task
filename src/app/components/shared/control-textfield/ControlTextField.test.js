import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

import ControlTextField from "./ControlTextField";
import { useForm } from "react-hook-form";

describe("test ControlTextField component", () => {
  const Component = () => {
    const { control } = useForm();
    return (
      <ControlTextField
        control={control}
        name="textfield"
        label="Control Textfield"
      />
    );
  };

  test("component is rendered correctly", () => {
    const component = renderer.create(<Component />).toJSON();
    expect(component).toMatchSnapshot();
  });

  test("check label text", () => {
    const { getByLabelText } = render(<Component />);
    const component = getByLabelText("Control Textfield");
    expect(component).toBeInTheDocument();
  });

  test("no helper text by default", () => {
    const { container } = render(<Component />);
    const component = container.getElementsByClassName(
      "MuiFormHelperText-root"
    );
    expect(component).toHaveLength(0);
  });
});
