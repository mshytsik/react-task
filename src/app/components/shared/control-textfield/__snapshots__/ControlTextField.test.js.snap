// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ControlTextField component component is rendered correctly 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth css-1lof2qr-MuiFormControl-root"
>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor=":r0:"
      id=":r0:-label"
    >
      Control Textfield
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        name="textfield"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            Control Textfield
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
